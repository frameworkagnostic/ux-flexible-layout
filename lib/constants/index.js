'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GRID12_GUTTERS = exports.GRID12_SIZES = exports.STEP = exports.UNIT = exports.BREAKPOINTS = exports.BREAKPOINT_KEYS = undefined;

var _breakpoint = require('./breakpoint');

Object.defineProperty(exports, 'BREAKPOINT_KEYS', {
  enumerable: true,
  get: function get() {
    return _breakpoint.BREAKPOINT_KEYS;
  }
});
Object.defineProperty(exports, 'BREAKPOINTS', {
  enumerable: true,
  get: function get() {
    return _breakpoint.BREAKPOINTS;
  }
});

var _grid12Column = require('./grid12Column');

var grid12Column = _interopRequireWildcard(_grid12Column);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var UNIT = exports.UNIT = 'px';
var STEP = exports.STEP = 5;

var GRID12_SIZES = exports.GRID12_SIZES = [].concat(_toConsumableArray(grid12Column.GRID_SIZES_DEFAULT), _toConsumableArray(grid12Column.GRID_SIZES));

var GRID12_GUTTERS = exports.GRID12_GUTTERS = grid12Column.GRID_GUTTERS;
//# sourceMappingURL=index.js.map