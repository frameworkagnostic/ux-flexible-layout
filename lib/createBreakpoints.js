'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keys = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.createBreakpoints = createBreakpoints;

var _index = require('./constants/index');

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } // Sorted ASC by size. That's important.
// It can't be configured as it's used statically for propTypes.

var keys = exports.keys = _index.BREAKPOINT_KEYS;

// Keep in mind that @media is inclusive by the CSS specification.
function createBreakpoints() {
  var breakpoints = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var _breakpoints$values = breakpoints.values,
      values = _breakpoints$values === undefined ? _index.BREAKPOINTS : _breakpoints$values,
      _breakpoints$unit = breakpoints.unit,
      unit = _breakpoints$unit === undefined ? _index.UNIT : _breakpoints$unit,
      _breakpoints$step = breakpoints.step,
      step = _breakpoints$step === undefined ? _index.STEP : _breakpoints$step,
      other = _objectWithoutProperties(breakpoints, ['values', 'unit', 'step']);

  function up(key) {
    var value = typeof values[key] === 'number' ? values[key] : key;
    return '@media (min-width:' + value + unit + ')';
  }

  function down(key) {
    var endIndex = keys.indexOf(key) + 1;
    var upperbound = values[keys[endIndex]];

    if (endIndex === keys.length) {
      // xl down applies to all sizes
      return up('xs');
    }

    var value = typeof upperbound === 'number' && endIndex > 0 ? upperbound : key;
    return '@media (max-width:' + (value - step) / 100 + unit + ')';
  }

  function between(start, end) {
    var endIndex = keys.indexOf(end) + 1;

    if (endIndex === keys.length) {
      return up(start);
    }

    return '@media (min-width:' + values[start] + unit + ') and ' + ('(max-width:' + (values[keys[endIndex]] - step) / 100 + unit + ')');
  }

  function only(key) {
    return between(key, key);
  }

  function width(key) {
    return values[key];
  }

  return _extends({
    keys: keys,
    values: values,
    up: up,
    down: down,
    between: between,
    only: only,
    width: width
  }, other);
}
//# sourceMappingURL=createBreakpoints.js.map