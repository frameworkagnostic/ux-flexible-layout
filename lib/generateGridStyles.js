'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.generateGrid = generateGrid;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function generateGrid(gridSizes, nextBreakpointUp, breakpoint) {
  var styles = {};

  gridSizes.forEach(function (size) {
    var key = 'grid-' + breakpoint + '-' + size;

    if (size === true) {
      // For the auto layouting
      styles[key] = {
        flexBasis: 0,
        flexGrow: 1,
        maxWidth: '100%'
      };
      return;
    }

    if (size === 'auto') {
      styles[key] = {
        flexBasis: 'auto',
        flexGrow: 0,
        maxWidth: 'none'
      };
      return;
    }

    // Only keep 6 significant numbers.
    var width = Math.round(size / 12 * 10e6) / 10e4 + '%';

    // Close to the bootstrap implementation:
    // https://github.com/twbs/bootstrap/blob/8fccaa2439e97ec72a4b7dc42ccc1f649790adb0/scss/mixins/_grid.scss#L41
    styles[key] = {
      flexBasis: width,
      flexGrow: 0,
      maxWidth: width
    };
  });

  return styles;
}

var generateGridReducer = exports.generateGridReducer = function generateGridReducer(gridSizes, nextBreakpointUp) {
  return function (accumulator, breakpoint) {
    var styles = generateGrid(gridSizes, nextBreakpointUp, breakpoint);

    if (breakpoint === 'xs') {
      return _extends({}, accumulator, styles);
    } else {
      return _extends({}, accumulator, _defineProperty({}, nextBreakpointUp(breakpoint), styles));
    }
  };
};
//# sourceMappingURL=generateGridStyles.js.map