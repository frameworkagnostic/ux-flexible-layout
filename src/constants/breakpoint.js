
export const BREAKPOINT_KEYS = [
  'xs',
  'sm',
  'md',
  'lg',
  'xl'
];

export const BREAKPOINTS = {
  xs: 0,
  sm: 600,
  md: 960,
  lg: 1280,
  xl: 1920,
};
