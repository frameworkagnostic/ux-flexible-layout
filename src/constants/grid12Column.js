
export const GRID_GUTTERS = [
  0,
  8,
  16,
  24,
  32,
  40
];

export const GRID_SIZES_DEFAULT = [
  'auto',
  true,
];

export const GRID_SIZES = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12
];
