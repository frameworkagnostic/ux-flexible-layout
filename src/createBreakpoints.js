// Sorted ASC by size. That's important.
// It can't be configured as it's used statically for propTypes.

import {
  BREAKPOINT_KEYS,
  BREAKPOINTS,
  UNIT,
  STEP
} from './constants/index';

export const keys = BREAKPOINT_KEYS;

// Keep in mind that @media is inclusive by the CSS specification.
export function createBreakpoints(breakpoints = {}) {
  const {
    // The breakpoint **start** at this value.
    // For instance with the first breakpoint xs: [xs, sm[.
    values = BREAKPOINTS,
    unit = UNIT,
    step = STEP,
    ...other
  } = breakpoints;

  function up(key) {
    const value = typeof values[key] === 'number' ? values[key] : key;
    return `@media (min-width:${value}${unit})`;
  }

  function down(key) {
    const endIndex = keys.indexOf(key) + 1;
    const upperbound = values[keys[endIndex]];

    if (endIndex === keys.length) {
      // xl down applies to all sizes
      return up('xs');
    }

    const value = typeof upperbound === 'number' && endIndex > 0 ? upperbound : key;
    return `@media (max-width:${(value - step) / 100}${unit})`;
  }

  function between(start, end) {
    const endIndex = keys.indexOf(end) + 1;

    if (endIndex === keys.length) {
      return up(start);
    }

    return (
      `@media (min-width:${values[start]}${unit}) and ` +
      `(max-width:${(values[keys[endIndex]] - step) / 100}${unit})`
    );
  }

  function only(key) {
    return between(key, key);
  }

  function width(key) {
    return values[key];
  }

  return {
    keys,
    values,
    up,
    down,
    between,
    only,
    width,
    ...other,
  };
}
