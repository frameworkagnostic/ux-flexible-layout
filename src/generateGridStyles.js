
export function generateGrid(gridSizes, nextBreakpointUp, breakpoint) {
  const styles = {};

  gridSizes.forEach((size) => {
    const key = `grid-${breakpoint}-${size}`;

    if (size === true) {
      // For the auto layouting
      styles[key] = {
        flexBasis: 0,
        flexGrow: 1,
        maxWidth: '100%',
      };
      return;
    }

    if (size === 'auto') {
      styles[key] = {
        flexBasis: 'auto',
        flexGrow: 0,
        maxWidth: 'none',
      };
      return;
    }

    // Only keep 6 significant numbers.
    const width = `${Math.round((size / 12) * 10e6) / 10e4}%`;

    // Close to the bootstrap implementation:
    // https://github.com/twbs/bootstrap/blob/8fccaa2439e97ec72a4b7dc42ccc1f649790adb0/scss/mixins/_grid.scss#L41
    styles[key] = {
      flexBasis: width,
      flexGrow: 0,
      maxWidth: width,
    };
  });

  return styles;
}

export const generateGridReducer = (gridSizes, nextBreakpointUp) => {
  return (accumulator, breakpoint) => {
    const styles = generateGrid(gridSizes, nextBreakpointUp, breakpoint);

    if (breakpoint === 'xs') {
      return {
        ...accumulator,
        ...styles,
      }
    } else {
      return {
        ...accumulator,
        ...{
          [nextBreakpointUp(breakpoint)]: styles
        },
      }
    }
  };
};
