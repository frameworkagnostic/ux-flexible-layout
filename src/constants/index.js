
import * as grid12Column from './grid12Column';

export {
  BREAKPOINT_KEYS,
  BREAKPOINTS,
} from './breakpoint';

export const UNIT = 'px';
export const STEP = 5;

export const GRID12_SIZES = [
  ...grid12Column.GRID_SIZES_DEFAULT,
  ...grid12Column.GRID_SIZES,
];

export const GRID12_GUTTERS = grid12Column.GRID_GUTTERS;
