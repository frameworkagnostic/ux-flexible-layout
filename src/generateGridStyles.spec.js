
import { keys, createBreakpoints } from './createBreakpoints';
import { generateGridReducer } from './generateGridStyles';
import { GRID12_SIZES } from './constants/index';

describe('generate grid', () => {
  test('isFunction', () => expect(generateGridReducer).toEqual(expect.any(Function)));

  describe('breakpoints', () => {
    test('snapshot', () => expect(keys).toMatchSnapshot());
  });

  describe('grid 12 columns', () => {
    test('snapshot', () => expect(GRID12_SIZES).toMatchSnapshot());

    describe('create styles', () => {
      const bp = createBreakpoints();
      const reducer = generateGridReducer(GRID12_SIZES, bp.up);
      const styles = keys.reduce(reducer, {});
      test('isObject', () => expect(styles).toEqual(expect.any(Object)));
      test('snapshot', () => expect(styles).toMatchSnapshot());
    });
  });
});
