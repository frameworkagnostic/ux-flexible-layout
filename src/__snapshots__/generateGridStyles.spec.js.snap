// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate grid breakpoints snapshot 1`] = `
Array [
  "xs",
  "sm",
  "md",
  "lg",
  "xl",
]
`;

exports[`generate grid grid 12 columns create styles snapshot 1`] = `
Object {
  "@media (min-width:1280px)": Object {
    "grid-lg-1": Object {
      "flexBasis": "8.33333%",
      "flexGrow": 0,
      "maxWidth": "8.33333%",
    },
    "grid-lg-10": Object {
      "flexBasis": "83.33333%",
      "flexGrow": 0,
      "maxWidth": "83.33333%",
    },
    "grid-lg-11": Object {
      "flexBasis": "91.66667%",
      "flexGrow": 0,
      "maxWidth": "91.66667%",
    },
    "grid-lg-12": Object {
      "flexBasis": "100%",
      "flexGrow": 0,
      "maxWidth": "100%",
    },
    "grid-lg-2": Object {
      "flexBasis": "16.66667%",
      "flexGrow": 0,
      "maxWidth": "16.66667%",
    },
    "grid-lg-3": Object {
      "flexBasis": "25%",
      "flexGrow": 0,
      "maxWidth": "25%",
    },
    "grid-lg-4": Object {
      "flexBasis": "33.33333%",
      "flexGrow": 0,
      "maxWidth": "33.33333%",
    },
    "grid-lg-5": Object {
      "flexBasis": "41.66667%",
      "flexGrow": 0,
      "maxWidth": "41.66667%",
    },
    "grid-lg-6": Object {
      "flexBasis": "50%",
      "flexGrow": 0,
      "maxWidth": "50%",
    },
    "grid-lg-7": Object {
      "flexBasis": "58.33333%",
      "flexGrow": 0,
      "maxWidth": "58.33333%",
    },
    "grid-lg-8": Object {
      "flexBasis": "66.66667%",
      "flexGrow": 0,
      "maxWidth": "66.66667%",
    },
    "grid-lg-9": Object {
      "flexBasis": "75%",
      "flexGrow": 0,
      "maxWidth": "75%",
    },
    "grid-lg-auto": Object {
      "flexBasis": "auto",
      "flexGrow": 0,
      "maxWidth": "none",
    },
    "grid-lg-true": Object {
      "flexBasis": 0,
      "flexGrow": 1,
      "maxWidth": "100%",
    },
  },
  "@media (min-width:1920px)": Object {
    "grid-xl-1": Object {
      "flexBasis": "8.33333%",
      "flexGrow": 0,
      "maxWidth": "8.33333%",
    },
    "grid-xl-10": Object {
      "flexBasis": "83.33333%",
      "flexGrow": 0,
      "maxWidth": "83.33333%",
    },
    "grid-xl-11": Object {
      "flexBasis": "91.66667%",
      "flexGrow": 0,
      "maxWidth": "91.66667%",
    },
    "grid-xl-12": Object {
      "flexBasis": "100%",
      "flexGrow": 0,
      "maxWidth": "100%",
    },
    "grid-xl-2": Object {
      "flexBasis": "16.66667%",
      "flexGrow": 0,
      "maxWidth": "16.66667%",
    },
    "grid-xl-3": Object {
      "flexBasis": "25%",
      "flexGrow": 0,
      "maxWidth": "25%",
    },
    "grid-xl-4": Object {
      "flexBasis": "33.33333%",
      "flexGrow": 0,
      "maxWidth": "33.33333%",
    },
    "grid-xl-5": Object {
      "flexBasis": "41.66667%",
      "flexGrow": 0,
      "maxWidth": "41.66667%",
    },
    "grid-xl-6": Object {
      "flexBasis": "50%",
      "flexGrow": 0,
      "maxWidth": "50%",
    },
    "grid-xl-7": Object {
      "flexBasis": "58.33333%",
      "flexGrow": 0,
      "maxWidth": "58.33333%",
    },
    "grid-xl-8": Object {
      "flexBasis": "66.66667%",
      "flexGrow": 0,
      "maxWidth": "66.66667%",
    },
    "grid-xl-9": Object {
      "flexBasis": "75%",
      "flexGrow": 0,
      "maxWidth": "75%",
    },
    "grid-xl-auto": Object {
      "flexBasis": "auto",
      "flexGrow": 0,
      "maxWidth": "none",
    },
    "grid-xl-true": Object {
      "flexBasis": 0,
      "flexGrow": 1,
      "maxWidth": "100%",
    },
  },
  "@media (min-width:600px)": Object {
    "grid-sm-1": Object {
      "flexBasis": "8.33333%",
      "flexGrow": 0,
      "maxWidth": "8.33333%",
    },
    "grid-sm-10": Object {
      "flexBasis": "83.33333%",
      "flexGrow": 0,
      "maxWidth": "83.33333%",
    },
    "grid-sm-11": Object {
      "flexBasis": "91.66667%",
      "flexGrow": 0,
      "maxWidth": "91.66667%",
    },
    "grid-sm-12": Object {
      "flexBasis": "100%",
      "flexGrow": 0,
      "maxWidth": "100%",
    },
    "grid-sm-2": Object {
      "flexBasis": "16.66667%",
      "flexGrow": 0,
      "maxWidth": "16.66667%",
    },
    "grid-sm-3": Object {
      "flexBasis": "25%",
      "flexGrow": 0,
      "maxWidth": "25%",
    },
    "grid-sm-4": Object {
      "flexBasis": "33.33333%",
      "flexGrow": 0,
      "maxWidth": "33.33333%",
    },
    "grid-sm-5": Object {
      "flexBasis": "41.66667%",
      "flexGrow": 0,
      "maxWidth": "41.66667%",
    },
    "grid-sm-6": Object {
      "flexBasis": "50%",
      "flexGrow": 0,
      "maxWidth": "50%",
    },
    "grid-sm-7": Object {
      "flexBasis": "58.33333%",
      "flexGrow": 0,
      "maxWidth": "58.33333%",
    },
    "grid-sm-8": Object {
      "flexBasis": "66.66667%",
      "flexGrow": 0,
      "maxWidth": "66.66667%",
    },
    "grid-sm-9": Object {
      "flexBasis": "75%",
      "flexGrow": 0,
      "maxWidth": "75%",
    },
    "grid-sm-auto": Object {
      "flexBasis": "auto",
      "flexGrow": 0,
      "maxWidth": "none",
    },
    "grid-sm-true": Object {
      "flexBasis": 0,
      "flexGrow": 1,
      "maxWidth": "100%",
    },
  },
  "@media (min-width:960px)": Object {
    "grid-md-1": Object {
      "flexBasis": "8.33333%",
      "flexGrow": 0,
      "maxWidth": "8.33333%",
    },
    "grid-md-10": Object {
      "flexBasis": "83.33333%",
      "flexGrow": 0,
      "maxWidth": "83.33333%",
    },
    "grid-md-11": Object {
      "flexBasis": "91.66667%",
      "flexGrow": 0,
      "maxWidth": "91.66667%",
    },
    "grid-md-12": Object {
      "flexBasis": "100%",
      "flexGrow": 0,
      "maxWidth": "100%",
    },
    "grid-md-2": Object {
      "flexBasis": "16.66667%",
      "flexGrow": 0,
      "maxWidth": "16.66667%",
    },
    "grid-md-3": Object {
      "flexBasis": "25%",
      "flexGrow": 0,
      "maxWidth": "25%",
    },
    "grid-md-4": Object {
      "flexBasis": "33.33333%",
      "flexGrow": 0,
      "maxWidth": "33.33333%",
    },
    "grid-md-5": Object {
      "flexBasis": "41.66667%",
      "flexGrow": 0,
      "maxWidth": "41.66667%",
    },
    "grid-md-6": Object {
      "flexBasis": "50%",
      "flexGrow": 0,
      "maxWidth": "50%",
    },
    "grid-md-7": Object {
      "flexBasis": "58.33333%",
      "flexGrow": 0,
      "maxWidth": "58.33333%",
    },
    "grid-md-8": Object {
      "flexBasis": "66.66667%",
      "flexGrow": 0,
      "maxWidth": "66.66667%",
    },
    "grid-md-9": Object {
      "flexBasis": "75%",
      "flexGrow": 0,
      "maxWidth": "75%",
    },
    "grid-md-auto": Object {
      "flexBasis": "auto",
      "flexGrow": 0,
      "maxWidth": "none",
    },
    "grid-md-true": Object {
      "flexBasis": 0,
      "flexGrow": 1,
      "maxWidth": "100%",
    },
  },
  "grid-xs-1": Object {
    "flexBasis": "8.33333%",
    "flexGrow": 0,
    "maxWidth": "8.33333%",
  },
  "grid-xs-10": Object {
    "flexBasis": "83.33333%",
    "flexGrow": 0,
    "maxWidth": "83.33333%",
  },
  "grid-xs-11": Object {
    "flexBasis": "91.66667%",
    "flexGrow": 0,
    "maxWidth": "91.66667%",
  },
  "grid-xs-12": Object {
    "flexBasis": "100%",
    "flexGrow": 0,
    "maxWidth": "100%",
  },
  "grid-xs-2": Object {
    "flexBasis": "16.66667%",
    "flexGrow": 0,
    "maxWidth": "16.66667%",
  },
  "grid-xs-3": Object {
    "flexBasis": "25%",
    "flexGrow": 0,
    "maxWidth": "25%",
  },
  "grid-xs-4": Object {
    "flexBasis": "33.33333%",
    "flexGrow": 0,
    "maxWidth": "33.33333%",
  },
  "grid-xs-5": Object {
    "flexBasis": "41.66667%",
    "flexGrow": 0,
    "maxWidth": "41.66667%",
  },
  "grid-xs-6": Object {
    "flexBasis": "50%",
    "flexGrow": 0,
    "maxWidth": "50%",
  },
  "grid-xs-7": Object {
    "flexBasis": "58.33333%",
    "flexGrow": 0,
    "maxWidth": "58.33333%",
  },
  "grid-xs-8": Object {
    "flexBasis": "66.66667%",
    "flexGrow": 0,
    "maxWidth": "66.66667%",
  },
  "grid-xs-9": Object {
    "flexBasis": "75%",
    "flexGrow": 0,
    "maxWidth": "75%",
  },
  "grid-xs-auto": Object {
    "flexBasis": "auto",
    "flexGrow": 0,
    "maxWidth": "none",
  },
  "grid-xs-true": Object {
    "flexBasis": 0,
    "flexGrow": 1,
    "maxWidth": "100%",
  },
}
`;

exports[`generate grid grid 12 columns snapshot 1`] = `
Array [
  "auto",
  true,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
]
`;
